File: frontend/lib/workspaces.ts
import { api } from "@/lib/api";
import type { AuthUser } from "@/lib/auth-token";

export type WorkspaceRole =
  | "owner"
  | "manager"
  | "analyst"
  | "viewer"
  | "platform_admin";

export type SellerWorkspace = {
  id: number;
  name: string;
  slug: string;
  status: "active" | "suspended" | "archived";
  current_role: WorkspaceRole;
  created_at: string;
  updated_at: string;
};

type TokenPayload = {
  access_token: string;
  token_type: string;
  user: AuthUser;
};

export type WorkspaceCreateResult = {
  workspace: SellerWorkspace;
  auth: TokenPayload;
};

export type WorkspaceMember = {
  user_id: number;
  email: string;
  name: string | null;
  role: Exclude<WorkspaceRole, "platform_admin">;
  joined_at: string;
};

export type MarketplaceShop = {
  id: number;
  workspace_id: number;
  platform: "shopee" | "lazada" | "tiktok_shop";
  external_shop_id: string;
  shop_name: string;
  status: "connected" | "expired" | "revoked" | "error";
  token_expires_at: string | null;
  last_synced_at: string | null;
  last_error: string | null;
  connected_at: string;
  updated_at: string;
};

export async function listWorkspaces(signal?: AbortSignal): Promise<SellerWorkspace[]> {
  const response = await api.get<SellerWorkspace[]>("/workspaces/", signal);
  return response.data ?? [];
}

export async function createWorkspace(name: string): Promise<WorkspaceCreateResult> {
  const response = await api.post<WorkspaceCreateResult>("/workspaces/", { name });
  if (!response.data) throw new Error("Workspace response is empty.");
  return response.data;
}

export async function listWorkspaceMembers(
  workspaceId: number,
  signal?: AbortSignal,
): Promise<WorkspaceMember[]> {
  const response = await api.get<WorkspaceMember[]>(
    `/workspaces/${workspaceId}/members`,
    signal,
  );
  return response.data ?? [];
}

export async function addWorkspaceMember(
  workspaceId: number,
  email: string,
  role: "manager" | "analyst" | "viewer",
): Promise<WorkspaceMember> {
  const response = await api.post<WorkspaceMember>(`/workspaces/${workspaceId}/members`, {
    email,
    role,
  });
  if (!response.data) throw new Error("Member response is empty.");
  return response.data;
}

export async function updateWorkspaceMemberRole(
  workspaceId: number,
  userId: number,
  role: WorkspaceMember["role"],
): Promise<WorkspaceMember> {
  const response = await api.patch<WorkspaceMember>(
    `/workspaces/${workspaceId}/members/${userId}`,
    { role },
  );
  if (!response.data) throw new Error("Member response is empty.");
  return response.data;
}

export async function removeWorkspaceMember(
  workspaceId: number,
  userId: number,
): Promise<void> {
  await api.delete(`/workspaces/${workspaceId}/members/${userId}`);
}

export async function listMarketplaceShops(
  workspaceId: number,
  signal?: AbortSignal,
): Promise<MarketplaceShop[]> {
  const response = await api.get<MarketplaceShop[]>(
    `/workspaces/${workspaceId}/shops`,
    signal,
  );
  return response.data ?? [];
}

export async function disconnectMarketplaceShop(
  workspaceId: number,
  shopId: number,
): Promise<MarketplaceShop> {
  const response = await api.delete<MarketplaceShop>(
    `/workspaces/${workspaceId}/shops/${shopId}`,
  );
  if (!response.data) throw new Error("Marketplace shop response is empty.");
  return response.data;
}

File: backend/app/api/v1/endpoints/workspaces.py
"""Authenticated seller-workspace onboarding and lookup."""

from __future__ import annotations

from typing import cast

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import (
    WorkspaceAccess,
    get_current_user,
    get_db_dep,
    get_workspace_access,
    require_workspace_role,
)
from app.core.exceptions import NotFoundError
from app.core.responses import ApiResponse, PageMeta
from app.schemas.workspace import (
    MarketplaceConnectionStatus,
    MarketplacePlatform,
    MarketplaceShopOut,
    WorkspaceCreateRequest,
    WorkspaceMemberAddRequest,
    WorkspaceMemberOut,
    WorkspaceMemberRoleRequest,
    WorkspaceOut,
    WorkspaceRole,
)
from app.services import marketplace_shop_service, user_service, workspace_service

router = APIRouter()
_OWNER_ACCESS = require_workspace_role("owner")
_SHOP_MANAGER_ACCESS = require_workspace_role("owner", "manager")


def _serialize(workspace, role: str) -> dict:  # noqa: ANN001
    return WorkspaceOut(
        id=workspace.id,
        name=workspace.name,
        slug=workspace.slug,
        status=workspace.status,
        current_role=cast(WorkspaceRole, role),
        created_at=workspace.created_at,
        updated_at=workspace.updated_at,
    ).model_dump()


def _serialize_member(membership, account) -> dict:  # noqa: ANN001
    return WorkspaceMemberOut(
        user_id=account.id,
        email=account.email,
        name=account.name,
        role=membership.role,
        joined_at=membership.created_at,
    ).model_dump()


def _serialize_shop(shop) -> dict:  # noqa: ANN001
    return MarketplaceShopOut(
        id=shop.id,
        workspace_id=shop.workspace_id,
        platform=cast(MarketplacePlatform, shop.platform),
        external_shop_id=shop.external_shop_id,
        shop_name=shop.shop_name,
        status=cast(MarketplaceConnectionStatus, shop.status),
        token_expires_at=shop.token_expires_at,
        last_synced_at=shop.last_synced_at,
        last_error=shop.last_error,
        connected_at=shop.created_at,
        updated_at=shop.updated_at,
    ).model_dump()


@router.post("/", response_model=ApiResponse[dict])
async def create_workspace(
    req: WorkspaceCreateRequest,
    user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db_dep),
) -> ApiResponse[dict]:
    workspace, membership, account = await workspace_service.create_workspace(
        db, user_id=int(user["sub"]), name=req.name, slug=req.slug
    )
    return ApiResponse[dict](
        success=True,
        data={
            "workspace": _serialize(workspace, membership.role),
            "auth": user_service.issue_token_response(account),
        },
        meta=PageMeta(),
        error=None,
    )


@router.get("/", response_model=ApiResponse[list[dict]])
async def list_workspaces(
    user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db_dep),
) -> ApiResponse[list[dict]]:
    rows = await workspace_service.list_accessible_workspaces(
        db, user_id=int(user["sub"]), is_platform_admin=user.get("role") == "admin"
    )
    items = [_serialize(workspace, role) for workspace, role in rows]
    return ApiResponse[list[dict]](
        success=True,
        data=items,
        meta=PageMeta(page=1, page_size=len(items), total=len(items)),
        error=None,
    )


@router.get("/{workspace_id}", response_model=ApiResponse[dict])
async def get_workspace(
    workspace_id: int,
    user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db_dep),
) -> ApiResponse[dict]:
    workspace, role = await workspace_service.get_accessible_workspace(
        db,
        workspace_id=workspace_id,
        user_id=int(user["sub"]),
        is_platform_admin=user.get("role") == "admin",
    )
    return ApiResponse[dict](
        success=True,
        data=_serialize(workspace, role),
        meta=PageMeta(),
        error=None,
    )


@router.get("/{workspace_id}/members", response_model=ApiResponse[list[dict]])
async def list_members(
    workspace_id: int,
    access: WorkspaceAccess = Depends(get_workspace_access),
    db: AsyncSession = Depends(get_db_dep),
) -> ApiResponse[list[dict]]:
    if access.workspace_id != workspace_id:
        # Header and path must describe one verified tenant; never silently use
        # one while showing the other in logs or browser history.
        raise NotFoundError("Không tìm thấy workspace.")
    rows = await workspace_service.list_members(db, workspace_id=workspace_id)
    items = [_serialize_member(membership, account) for membership, account in rows]
    return ApiResponse[list[dict]](
        success=True,
        data=items,
        meta=PageMeta(page=1, page_size=len(items), total=len(items)),
        error=None,
    )


@router.post("/{workspace_id}/members", response_model=ApiResponse[dict])
async def add_member(
    workspace_id: int,
    req: WorkspaceMemberAddRequest,
    access: WorkspaceAccess = Depends(_OWNER_ACCESS),
    db: AsyncSession = Depends(get_db_dep),
) -> ApiResponse[dict]:
    if access.workspace_id != workspace_id:
        raise NotFoundError("Không tìm thấy workspace.")
    membership, account = await workspace_service.add_member(
        db, workspace_id=workspace_id, email=req.email, role=req.role
    )
    return ApiResponse[dict](
        success=True,
        data=_serialize_member(membership, account),
        meta=PageMeta(),
        error=None,
    )


@router.patch("/{workspace_id}/members/{member_user_id}", response_model=ApiResponse[dict])
async def update_member_role(
    workspace_id: int,
    member_user_id: int,
    req: WorkspaceMemberRoleRequest,
    access: WorkspaceAccess = Depends(_OWNER_ACCESS),
    db: AsyncSession = Depends(get_db_dep),
) -> ApiResponse[dict]:
    if access.workspace_id != workspace_id:
        raise NotFoundError("Không tìm thấy workspace.")
    membership, account = await workspace_service.update_member_role(
        db,
        workspace_id=workspace_id,
        member_user_id=member_user_id,
        role=req.role,
    )
    return ApiResponse[dict](
        success=True,
        data=_serialize_member(membership, account),
        meta=PageMeta(),
        error=None,
    )


@router.delete("/{workspace_id}/members/{member_user_id}", response_model=ApiResponse[dict])
async def remove_member(
    workspace_id: int,
    member_user_id: int,
    access: WorkspaceAccess = Depends(_OWNER_ACCESS),
    db: AsyncSession = Depends(get_db_dep),
) -> ApiResponse[dict]:
    if access.workspace_id != workspace_id:
        raise NotFoundError("Không tìm thấy workspace.")
    await workspace_service.remove_member(
        db,
        workspace_id=workspace_id,
        member_user_id=member_user_id,
        acting_user_id=access.user_id,
    )
    return ApiResponse[dict](
        success=True,
        data={"removed": True},
        meta=PageMeta(),
        error=None,
    )


@router.get("/{workspace_id}/shops", response_model=ApiResponse[list[dict]])
async def list_connected_shops(
    workspace_id: int,
    access: WorkspaceAccess = Depends(get_workspace_access),
    db: AsyncSession = Depends(get_db_dep),
) -> ApiResponse[list[dict]]:
    if access.workspace_id != workspace_id:
        raise NotFoundError("Không tìm thấy workspace.")
    shops = await marketplace_shop_service.list_shops(db, workspace_id=workspace_id)
    items = [_serialize_shop(shop) for shop in shops]
    return ApiResponse[list[dict]](
        success=True,
        data=items,
        meta=PageMeta(page=1, page_size=len(items), total=len(items)),
        error=None,
    )


@router.delete("/{workspace_id}/shops/{shop_id}", response_model=ApiResponse[dict])
async def disconnect_shop(
    workspace_id: int,
    shop_id: int,
    access: WorkspaceAccess = Depends(_SHOP_MANAGER_ACCESS),
    db: AsyncSession = Depends(get_db_dep),
) -> ApiResponse[dict]:
    if access.workspace_id != workspace_id:
        raise NotFoundError("Không tìm thấy workspace.")
    shop = await marketplace_shop_service.disconnect_shop(
        db, workspace_id=workspace_id, shop_id=shop_id
    )
    return ApiResponse[dict](
        success=True,
        data=_serialize_shop(shop),
        meta=PageMeta(),
        error=None,
    )
